File: src/main/java/com/ultimate/component/TableInfo.java
package com.ultimate.component;

import com.ultimate.component.info.ComponentInfo;
import com.ultimate.exception.ComponentNotExistsException;

import java.util.*;

public class TableInfo{

    private static Map<String, ComponentInfo> TypeMap;
    private static List<ComponentInfo> componentList;

    public static void putComponent(Class clazz, ComponentInfo componentInfo){

        Objects.requireNonNull(componentInfo, componentInfo.getTableName());
        getTypeMapInstance().put(clazz.getName(), componentInfo);
        getTableNameListInstance().add(componentInfo);

    }

    public static ComponentInfo getComponent(Class clazz){

        if(clazz == null){
            throw new NullPointerException("param clazz must not null!");
        }

        ComponentInfo componentInfo = getTypeMapInstance().get(clazz.getName());
        if(componentInfo==null){
            throw new ComponentNotExistsException(clazz+" component is not exists!");
        }

        return componentInfo;
    }

    public static String getTableName(Class clazz){

        if(clazz == null){
            throw new NullPointerException("param calzz must not null!");
        }

        ComponentInfo componentInfo = getTypeMapInstance().get(clazz.getName());
        if(componentInfo != null){
            return componentInfo.getTableName();
        }

        return null;
    }

    public static List<ComponentInfo> getComponentList(){

        return componentList;
    }

    private static Map<String, ComponentInfo> getTypeMapInstance(){

        if(TypeMap == null){
            TypeMap = new HashMap<>();
        }

        return TypeMap;
    }

    private static List<ComponentInfo> getTableNameListInstance(){

        if(componentList == null){
            componentList = new ArrayList<>();
        }

        return componentList;
    }

}

File: src/main/java/github/cweijan/ultimate/core/component/TableInfo.kt
package github.cweijan.ultimate.core.component

import github.cweijan.ultimate.core.component.info.ComponentInfo
import github.cweijan.ultimate.exception.ComponentNotExistsException

object TableInfo {

    private var developMode = false;

    fun enableDevelopMode() {
        developMode = true
    }

    private val TypeMap: MutableMap<String?, ComponentInfo> by lazy {
        return@lazy HashMap<String?, ComponentInfo>()
    }

    val componentList: MutableList<ComponentInfo> by lazy {
        return@lazy ArrayList<ComponentInfo>()
    }

    @JvmStatic
    fun putComponent(clazz: Class<*>, componentInfo: ComponentInfo) {
        TypeMap[clazz.name] = componentInfo
        componentList.add(componentInfo)
    }

    @JvmStatic
    fun removeComponent(clazz: Class<*>?) {
        clazz ?: return
        componentList.remove(TypeMap[clazz.name])
        TypeMap.remove(clazz.name)
    }

    fun isAlreadyInit(clazz: Class<*>): Boolean {
        return TypeMap.containsKey(clazz.name)
    }


    fun getComponent(clazz: Class<*>?, nullable: Boolean = false): ComponentInfo? {

        if(developMode){
            clazz?.let { ComponentInfo.init(it,false) }
        }

        val componentInfo = TypeMap[clazz?.name] ?: clazz?.let { ComponentInfo.init(it) }
        return componentInfo
                ?: if (nullable) return null else throw ComponentNotExistsException("$clazz component is not exists!")
    }

    @JvmStatic
    fun getComponent(clazz: Class<*>?): ComponentInfo {

        return getComponent(clazz, false)!!
    }

    fun getTableName(clazz: Class<*>): String? {
        return getComponent(clazz).tableName;
    }

}
